app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,74 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
+  map: Map<string, WeakRefCompatibleObject<Record<string, object>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
+    Object.keys(value).forEach((valueKey) => {
+      weakReffedValueObj[valueKey] = new WeakRef(value[valueKey]);
+    });
+    this.map.set(key, weakReffedValueObj);
+  }
+
+  get(key: string) {
+    const weakReffedValue = this.map.get(key);
+    if (!weakReffedValue) {
+      return undefined;
+    }
+
+    const deReffedValue = {} as { [dereffedValueObjKey: string]: WeakKey };
```

Comment:
Nit: A type annotation would be preferred here as well, for the same reason given in the other comment:

```suggestion
    const deReffedValue: { [dereffedValueObjKey: string]: WeakKey } = {};
```

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646